internal/core/src/common/Array.h

```diff
@@ -0,0 +1,366 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <utility>
+#include <vector>
+#include <arrow/array.h>
+#include <arrow/array/builder_primitive.h>
+#include "fmt/core.h"
+#include "exceptions/EasyAssert.h"
+#include "Types.h"
+#include "FieldMeta.h"
+
+namespace milvus {
+
+class Array {
+ public:
+    Array() = default;
+
+    explicit Array(const ScalarArray& field_data) {
+        switch (field_data.data_case()) {
+            case ScalarArray::kBoolData: {
+                element_type_ = DataType::BOOL;
+                length_ = field_data.bool_data().data().size();
+                data_ = new char[length_ * sizeof(bool)];
+                for (int i = 0; i < length_; ++i) {
+                    reinterpret_cast<bool*>(data_)[i] =
+                        field_data.bool_data().data(i);
+                    offsets_.emplace_back(size_);
+                    size_ += sizeof(bool);
+                }
+                break;
+            }
+            case ScalarArray::kIntData: {
+                element_type_ = DataType::INT32;
+                length_ = field_data.int_data().data().size();
+                data_ = new char[length_ * sizeof(int)];
+                for (int i = 0; i < length_; ++i) {
+                    reinterpret_cast<int*>(data_)[i] =
+                        field_data.int_data().data(i);
+                    offsets_.emplace_back(size_);
```

Comment:
vector may cost copy and relayout memory when dynamic increase, for int/float/double type:
can rewrite:
....
size_ = sizeof(T) * length_;
offsets.reserve(length_);
for(size_t i = 0; i < length_; ++i ) {
...
offsets.push_back(sizeof(T) *i);
}